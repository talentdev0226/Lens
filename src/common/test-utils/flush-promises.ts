/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { setImmediate } from "timers";

export const flushPromises = () => new Promise(setImmediate);
