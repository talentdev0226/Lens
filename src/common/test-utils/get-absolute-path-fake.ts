/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { GetAbsolutePath } from "../path/get-absolute-path.injectable";

export const getAbsolutePathFake: GetAbsolutePath = (...args) => {
  const maybeAbsolutePath = args.join("/");

  if (isAbsolutePath(maybeAbsolutePath)) {
    return maybeAbsolutePath;
  }

  return `/some-absolute-root-directory/${maybeAbsolutePath}`;
};

const isAbsolutePath = (path: string) => path.startsWith("/");
