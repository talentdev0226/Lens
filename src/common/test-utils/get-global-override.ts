/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { Injectable } from "@ogre-tools/injectable";

export interface GlobalOverride {
  injectable: Injectable<any, any, any>;
  overridingInstantiate: any;
}

export const getGlobalOverride = <T extends Injectable<any, any, any>>(
  injectable: T,
  overridingInstantiate: T["instantiate"],
) => ({
    injectable,
    overridingInstantiate,
  });
