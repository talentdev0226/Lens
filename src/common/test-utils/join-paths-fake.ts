/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { JoinPaths } from "../path/join-paths.injectable";

export const joinPathsFake: JoinPaths = (...args) => args.join("/");
