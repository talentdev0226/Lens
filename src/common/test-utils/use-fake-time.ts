/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { act } from "@testing-library/react";

let usingFakeTime = false;

export const advanceFakeTime = (milliseconds: number) => {
  if (!usingFakeTime) {
    throw new Error("Tried to advance fake time but it was not enabled. Call useFakeTime() first.");
  }

  act(() => {
    jest.advanceTimersByTime(milliseconds);
  });
};

export const useFakeTime = (dateTime = "2015-10-21T07:28:00Z") => {
  usingFakeTime = true;

  jest.useFakeTimers();

  jest.setSystemTime(new Date(dateTime));
};
