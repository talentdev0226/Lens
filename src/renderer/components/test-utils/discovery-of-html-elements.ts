/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { RenderResult } from "@testing-library/react";

type DiscoverySourceTypes = RenderResult | Element;

export type QuerySingleElement = (
  attributeName: string,
  attributeValue?: string
) => { discovered: Element | null } & Discover;

export type GetSingleElement = (
  attributeName: string,
  attributeValue?: string
) => { discovered: Element } & Discover;

export type QueryAllElements = (attributeName: string) => {
  discovered: Element[];
  attributeValues: (string | null)[];
};

export interface Discover {
  querySingleElement: QuerySingleElement;
  queryAllElements: QueryAllElements;
  getSingleElement: GetSingleElement;
}

export const discoverFor = (getSource: () => DiscoverySourceTypes): Discover => ({
  querySingleElement: querySingleElement(getSource),
  queryAllElements: queryAllElements(getSource),
  getSingleElement: getSingleElement(getSource),
});

export const querySingleElement =
  (getSource: () => DiscoverySourceTypes): QuerySingleElement =>
    (attributeName, attributeValue) => {
      const source = getSource();

      const dataAttribute = `data-${attributeName}-test`;

      const selector = attributeValue
        ? `[${dataAttribute}="${attributeValue}"]`
        : `[${dataAttribute}]`;

      const discovered = getBaseElement(source).querySelector(selector);

      const nestedDiscover = discoverFor(() => {
        if (!discovered) {
          throw new Error("Tried to do nested discover using source that does not exist");
        }

        return discovered;
      });

      return {
        discovered,

        ...nestedDiscover,
      };
    };

export const queryAllElements =
  (getSource: () => DiscoverySourceTypes): QueryAllElements =>
    (attributeName) => {
      const source = getSource();

      const dataAttribute = `data-${attributeName}-test`;

      const results = [
        ...getBaseElement(source).querySelectorAll(`[${dataAttribute}]`),
      ];

      return {
        discovered: results,

        attributeValues: results.map((result) =>
          result.getAttribute(dataAttribute),
        ),
      };
    };

export const getSingleElement =
  (getSource: () => DiscoverySourceTypes): GetSingleElement =>
    (attributeName, attributeValue) => {
      const dataAttribute = `data-${attributeName}-test`;

      const { discovered, ...nestedDiscover } = querySingleElement(getSource)(
        attributeName,
        attributeValue,
      );

      if (!discovered) {
        const validValues =
        queryAllElements(getSource)(attributeName).attributeValues;

        if (attributeValue) {
          throw new Error(
            `Couldn't find HTML-element with attribute "${dataAttribute}" with value "${attributeValue}". Present values are:\n\n"${validValues.join(
              '",\n"',
            )}"`,
          );
        }

        throw new Error(
          `Couldn't find HTML-element with attribute "${dataAttribute}"`,
        );
      }

      return { discovered, ...nestedDiscover };
    };

const getBaseElement = (source: DiscoverySourceTypes) =>
  "baseElement" in source ? source.baseElement : source;
