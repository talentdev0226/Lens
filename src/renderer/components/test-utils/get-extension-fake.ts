/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { Mutable } from "type-fest";
import fileSystemProvisionerStoreInjectable from "../../../extensions/extension-loader/file-system-provisioner-store/file-system-provisioner-store.injectable";
import { lensExtensionDependencies } from "../../../extensions/lens-extension";
import { LensMainExtension } from "../../../extensions/lens-main-extension";
import navigateForExtensionInjectable from "../../../main/start-main-application/lens-window/navigate-for-extension.injectable";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import catalogCategoryRegistryInjectable from "../../../common/catalog/category-registry.injectable";
import getExtensionPageParametersInjectable from "../../routes/get-extension-page-parameters.injectable";
import navigateToRouteInjectable from "../../routes/navigate-to-route.injectable";
import routesInjectable from "../../routes/routes.injectable";
import catalogEntityRegistryForMainInjectable from "../../../main/catalog/entity-registry.injectable";
import catalogEntityRegistryForRendererInjectable from "../../api/catalog/entity/registry.injectable";
import type { DiContainer } from "@ogre-tools/injectable";

export class TestExtensionMain extends LensMainExtension {}
export class TestExtensionRenderer extends LensRendererExtension {}

export interface FakeExtensionOptions {
  id: string;
  name: string;
  rendererOptions?: Partial<LensRendererExtension>;
  mainOptions?: Partial<LensMainExtension>;
}

export const getExtensionFakeForMain = (di: DiContainer, id: string, name: string, options: Partial<LensMainExtension>) => {
  const instance = new TestExtensionMain({
    id,
    absolutePath: "irrelevant",
    isBundled: false,
    isCompatible: false,
    isEnabled: false,
    manifest: {
      name,
      version: "1.0.0",
      engines: {
        lens: "^5.5.0",
      },
    },
    manifestPath: "irrelevant",
  });

  Object.assign(instance, options);

  (instance as Mutable<LensMainExtension>)[lensExtensionDependencies] = {
    fileSystemProvisionerStore: di.inject(
      fileSystemProvisionerStoreInjectable,
    ),
    entityRegistry: di.inject(catalogEntityRegistryForMainInjectable),
    navigate: di.inject(navigateForExtensionInjectable),
  };

  return instance;
};

export const getExtensionFakeForRenderer = (di: DiContainer, id: string, name: string, options: Partial<LensRendererExtension>) => {
  const instance = new TestExtensionRenderer({
    id,
    absolutePath: "irrelevant",
    isBundled: false,
    isCompatible: false,
    isEnabled: false,
    manifest: {
      name,
      version: "1.0.0",
      engines: {
        lens: "^5.5.0",
      },
    },
    manifestPath: "irrelevant",
  });

  Object.assign(instance, options);

  (instance as Mutable<LensRendererExtension>)[lensExtensionDependencies] = {
    categoryRegistry: di.inject(catalogCategoryRegistryInjectable),
    entityRegistry: di.inject(catalogEntityRegistryForRendererInjectable),
    fileSystemProvisionerStore: di.inject(fileSystemProvisionerStoreInjectable),
    getExtensionPageParameters: di.inject(getExtensionPageParametersInjectable),
    navigateToRoute: di.inject(navigateToRouteInjectable),
    routes: di.inject(routesInjectable),
  };

  return instance;
};
