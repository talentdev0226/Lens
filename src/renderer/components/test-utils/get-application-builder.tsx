/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed, IObservableArray, observable, runInAction } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../../common/cluster-store/allowed-resources.injectable";
import { fireEvent, RenderResult } from "@testing-library/react";
import type { KubeResource } from "../../../common/rbac";
import directoryForLensLocalStorageInjectable from "../../../common/directory-for-lens-local-storage/directory-for-lens-local-storage.injectable";
import { Sidebar } from "../layout/sidebar";
import { getDisForUnitTesting } from "../../../test-utils/get-dis-for-unit-testing";
import type { DiContainer } from "@ogre-tools/injectable";
import clusterStoreInjectable from "../../../common/cluster-store/cluster-store.injectable";
import type { ClusterStore } from "../../../common/cluster-store/cluster-store";
import mainExtensionsInjectable from "../../../extensions/main-extensions.injectable";
import type { LensMainExtension } from "../../../extensions/lens-main-extension";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { pipeline } from "@ogre-tools/fp";
import { flatMap, compact, join, get, filter } from "lodash/fp";
import preferenceNavigationItemsInjectable from "../+preferences/preferences-navigation/preference-navigation-items.injectable";
import navigateToPreferencesInjectable from "../../../common/front-end-routing/routes/preferences/navigate-to-preferences.injectable";
import applicationMenuItemsInjectable, { MenuItemOpts } from "../../../main/menu/application-menu-items.injectable";

type Callback = (dis: DiContainers) => void | Promise<void>;

export interface ApplicationBuilder {
  dis: DiContainers;
  setEnvironmentToClusterFrame: () => ApplicationBuilder;
  addExtensions: (...extensions: LensRendererExtension[]) => Promise<ApplicationBuilder>;
  allowKubeResource: (resourceName: KubeResource) => ApplicationBuilder;
  beforeSetups: (callback: Callback) => ApplicationBuilder;
  beforeRender: (callback: Callback) => ApplicationBuilder;
  render: () => Promise<RenderResult>;

  applicationMenu: {
    click: (path: string) => void;
  };

  preferences: {
    close: () => void;
    navigate: () => void;
    navigation: {
      click: (id: string) => void;
    };
  };
}

interface DiContainers {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}

interface Environment {
  renderSidebar: () => React.ReactNode;
  onAllowKubeResource: () => void;
}

export const getApplicationBuilder = () => {
  const { rendererDi, mainDi, runSetups } = getDisForUnitTesting({
    doGeneralOverrides: true,
  });

  const dis = { rendererDi, mainDi };

  const clusterStoreStub = {
    getById: (): null => null,
  } as unknown as ClusterStore;

  rendererDi.override(clusterStoreInjectable, () => clusterStoreStub);
  mainDi.override(clusterStoreInjectable, () => clusterStoreStub);

  const beforeSetupsCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  const extensionsState = observable.array<LensRendererExtension>();

  rendererDi.override(subscribeStoresInjectable, () => () => () => {});

  const environments: Record<string, Environment> = {
    application: {
      renderSidebar: () => null,

      onAllowKubeResource: () => {
        throw new Error(
          "Tried to allow kube resource when environment is not cluster frame.",
        );
      },
    },

    clusterFrame: {
      renderSidebar: () => <Sidebar />,
      onAllowKubeResource: () => {},
    },
  };

  let environment = environments.application;

  rendererDi.override(
    currentlyInClusterFrameInjectable,
    () => environment === environments.clusterFrame,
  );

  rendererDi.override(rendererExtensionsInjectable, () =>
    computed(() => extensionsState),
  );

  mainDi.override(mainExtensionsInjectable, () =>
    computed((): LensMainExtension[] => []),
  );

  let allowedResourcesState: IObservableArray<KubeResource>;
  let rendered: RenderResult;

  const builder: ApplicationBuilder = {
    dis,

    applicationMenu: {
      click: (path: string) => {
        const applicationMenuItems = mainDi.inject(
          applicationMenuItemsInjectable,
        );

        const menuItems = pipeline(
          applicationMenuItems.get(),
          flatMap(toFlatChildren(null)),
          filter((menuItem) => !!menuItem.click),
        );

        const menuItem = menuItems.find((menuItem) => menuItem.path === path);

        if (!menuItem) {
          const availableIds = menuItems.map(get("path")).join('", "');

          throw new Error(
            `Tried to click application menu item with ID "${path}" which does not exist. Available IDs are: "${availableIds}"`,
          );
        }

        menuItem.click(undefined, undefined, undefined);
      },
    },

    preferences: {
      close: () => {
        const link = rendered.getByTestId("close-preferences");

        fireEvent.click(link);
      },

      navigate: () => {
        const navigateToPreferences = rendererDi.inject(navigateToPreferencesInjectable);

        navigateToPreferences();
      },

      navigation: {
        click: (id: string) => {
          const link = rendered.queryByTestId(`tab-link-for-${id}`);

          if (!link) {
            const preferencesNavigationItems = rendererDi.inject(
              preferenceNavigationItemsInjectable,
            );

            const availableIds = preferencesNavigationItems
              .get()
              .map(get("id"));

            throw new Error(
              `Tried to click navigation item "${id}" which does not exist in preferences. Available IDs are "${availableIds.join(
                '", "',
              )}"`,
            );
          }

          fireEvent.click(link);
        },
      },
    },

    setEnvironmentToClusterFrame: () => {
      environment = environments.clusterFrame;

      allowedResourcesState = observable.array();

      rendererDi.override(allowedResourcesInjectable, () =>
        computed(() => new Set([...allowedResourcesState])),
      );

      rendererDi.override(
        directoryForLensLocalStorageInjectable,
        () => "/irrelevant",
      );

      return builder;
    },

    addExtensions: async (...extensions) => {
      const extensionRegistrators = rendererDi.injectMany(
        extensionRegistratorInjectionToken,
      );

      const addAndEnableExtensions = async () => {
        const registratorPromises = extensions.flatMap((extension) =>
          extensionRegistrators.map((registrator) => registrator(extension, 1)),
        );

        await Promise.all(registratorPromises);

        runInAction(() => {
          extensions.forEach((extension) => {
            extensionsState.push(extension);
          });
        });
      };

      if (rendered) {
        await addAndEnableExtensions();
      } else {
        builder.beforeRender(addAndEnableExtensions);
      }

      return builder;
    },

    allowKubeResource: (resourceName) => {
      environment.onAllowKubeResource();

      runInAction(() => {
        allowedResourcesState.push(resourceName);
      });

      return builder;
    },

    beforeSetups(callback: (dis: DiContainers) => void) {
      beforeSetupsCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: (dis: DiContainers) => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      for (const callback of beforeSetupsCallbacks) {
        await callback(dis);
      }

      await runSetups();

      const render = renderFor(rendererDi);

      const history = rendererDi.inject(observableHistoryInjectable);

      const currentRouteComponent = rendererDi.inject(
        currentRouteComponentInjectable,
      );

      for (const callback of beforeRenderCallbacks) {
        await callback(dis);
      }

      rendered = render(
        <Router history={history}>
          {environment.renderSidebar()}

          <Observer>
            {() => {
              const Component = currentRouteComponent.get();

              if (!Component) {
                return null;
              }

              return <Component />;
            }}
          </Observer>
        </Router>,
      );

      return rendered;
    },
  };

  return builder;
};

const toFlatChildren =
  (parentId: string) =>
    ({
      submenu = [],
      ...menuItem
    }: MenuItemOpts): (MenuItemOpts & { path: string })[] =>
      [
        {
          ...menuItem,
          path: pipeline([parentId, menuItem.id], compact, join(".")),
        },
        ...submenu.flatMap(toFlatChildren(menuItem.id)),
      ];
