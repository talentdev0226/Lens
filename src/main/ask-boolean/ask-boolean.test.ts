/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { ApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import { getApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import type { AskBoolean } from "./ask-boolean.injectable";
import askBooleanInjectable from "./ask-boolean.injectable";
import { getPromiseStatus } from "../../common/test-utils/get-promise-status";
import type { RenderResult } from "@testing-library/react";
import { fireEvent } from "@testing-library/react";
import getRandomIdInjectable from "../../common/utils/get-random-id.injectable";

describe("ask-boolean", () => {
  let applicationBuilder: ApplicationBuilder;
  let askBoolean: AskBoolean;

  beforeEach(() => {
    applicationBuilder = getApplicationBuilder();

    const getRandomIdFake = jest
      .fn()
      .mockReturnValueOnce("some-random-id-1")
      .mockReturnValueOnce("some-random-id-2");

    applicationBuilder.dis.mainDi.override(getRandomIdInjectable, () => getRandomIdFake);

    askBoolean = applicationBuilder.dis.mainDi.inject(askBooleanInjectable);
  });

  describe("given started", () => {
    let rendered: RenderResult;

    beforeEach(async () => {
      rendered = await applicationBuilder.render();
    });

    describe("when asking question", () => {
      let actualPromise: Promise<boolean>;

      beforeEach(() => {
        actualPromise = askBoolean({
          title: "some-title",
          question: "Some question",
        });
      });

      it("does not resolve yet", async () => {
        const promiseStatus = await getPromiseStatus(actualPromise);

        expect(promiseStatus.fulfilled).toBe(false);
      });

      it("renders", () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      it("shows notification", () => {
        const notification = rendered.getByTestId("ask-boolean-some-random-id-1");

        expect(notification).not.toBeUndefined();
      });

      describe('when user answers "yes"', () => {
        beforeEach(() => {
          const button = rendered.getByTestId("ask-boolean-some-random-id-1-button-yes");

          fireEvent.click(button);
        });

        it("renders", () => {
          expect(rendered.baseElement).toMatchSnapshot();
        });

        it("does not show notification anymore", () => {
          const notification = rendered.queryByTestId("ask-boolean-some-random-id-1");

          expect(notification).toBeNull();
        });

        it("resolves", async () => {
          const actual = await actualPromise;

          expect(actual).toBe(true);
        });
      });

      describe('when user answers "no"', () => {
        beforeEach(() => {
          const button = rendered.getByTestId("ask-boolean-some-random-id-1-button-no");

          fireEvent.click(button);
        });

        it("renders", () => {
          expect(rendered.baseElement).toMatchSnapshot();
        });

        it("does not show notification anymore", () => {
          const notification = rendered.queryByTestId("ask-boolean-some-random-id-1");

          expect(notification).toBeNull();
        });

        it("resolves", async () => {
          const actual = await actualPromise;

          expect(actual).toBe(false);
        });
      });

      describe("when user closes notification without answering the question", () => {
        beforeEach(() => {
          const button = rendered.getByTestId("close-notification-for-ask-boolean-for-some-random-id-1");

          fireEvent.click(button);
        });

        it("renders", () => {
          expect(rendered.baseElement).toMatchSnapshot();
        });

        it("does not show notification anymore", () => {
          const notification = rendered.queryByTestId("ask-boolean-some-random-id-1");

          expect(notification).toBeNull();
        });

        it("resolves", async () => {
          const actual = await actualPromise;

          expect(actual).toBe(false);
        });
      });
    });

    describe("when asking multiple questions", () => {
      let firstQuestionPromise: Promise<boolean>;
      let secondQuestionPromise: Promise<boolean>;

      beforeEach(() => {
        firstQuestionPromise = askBoolean({
          title: "some-title",
          question: "Some question",
        });

        secondQuestionPromise = askBoolean({
          title: "some-other-title",
          question: "Some other question",
        });
      });

      it("renders", () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      it("shows notification for first question", () => {
        const notification = rendered.getByTestId("ask-boolean-some-random-id-1");

        expect(notification).not.toBeUndefined();
      });

      it("shows notification for second question", () => {
        const notification = rendered.getByTestId("ask-boolean-some-random-id-2");

        expect(notification).not.toBeUndefined();
      });

      describe("when answering to first question", () => {
        beforeEach(() => {
          const button = rendered.getByTestId("ask-boolean-some-random-id-1-button-no");

          fireEvent.click(button);
        });

        it("renders", () => {
          expect(rendered.baseElement).toMatchSnapshot();
        });

        it("does not show notification for first question anymore", () => {
          const notification = rendered.queryByTestId("ask-boolean-some-random-id-1");

          expect(notification).toBeNull();
        });

        it("still shows notification for second question", () => {
          const notification = rendered.getByTestId("ask-boolean-some-random-id-2");

          expect(notification).not.toBeUndefined();
        });

        it("resolves first question", async () => {
          const actual = await firstQuestionPromise;

          expect(actual).toBe(false);
        });

        it("does not resolve second question yet", async () => {
          const promiseStatus = await getPromiseStatus(secondQuestionPromise);

          expect(promiseStatus.fulfilled).toBe(false);
        });
      });
    });
  });
});
