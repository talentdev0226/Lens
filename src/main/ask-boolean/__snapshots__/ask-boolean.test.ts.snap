// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ask-boolean given started when asking multiple questions renders 1`] = `
<body>
  <div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          <div
            class="flex column gaps"
            data-testid="ask-boolean-some-random-id-1"
          >
            <b>
              some-title
            </b>
            <p>
              Some question
            </p>
            <div
              class="flex gaps row align-left box grow"
            >
              <button
                class="Button light"
                data-testid="ask-boolean-some-random-id-1-button-yes"
                type="button"
              >
                Yes
              </button>
              <button
                class="Button active outlined"
                data-testid="ask-boolean-some-random-id-1-button-no"
                type="button"
              >
                No
              </button>
            </div>
          </div>
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-ask-boolean-for-some-random-id-1"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          <div
            class="flex column gaps"
            data-testid="ask-boolean-some-random-id-2"
          >
            <b>
              some-other-title
            </b>
            <p>
              Some other question
            </p>
            <div
              class="flex gaps row align-left box grow"
            >
              <button
                class="Button light"
                data-testid="ask-boolean-some-random-id-2-button-yes"
                type="button"
              >
                Yes
              </button>
              <button
                class="Button active outlined"
                data-testid="ask-boolean-some-random-id-2-button-no"
                type="button"
              >
                No
              </button>
            </div>
          </div>
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-ask-boolean-for-some-random-id-2"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ask-boolean given started when asking multiple questions when answering to first question renders 1`] = `
<body>
  <div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          <div
            class="flex column gaps"
            data-testid="ask-boolean-some-random-id-2"
          >
            <b>
              some-other-title
            </b>
            <p>
              Some other question
            </p>
            <div
              class="flex gaps row align-left box grow"
            >
              <button
                class="Button light"
                data-testid="ask-boolean-some-random-id-2-button-yes"
                type="button"
              >
                Yes
              </button>
              <button
                class="Button active outlined"
                data-testid="ask-boolean-some-random-id-2-button-no"
                type="button"
              >
                No
              </button>
            </div>
          </div>
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-ask-boolean-for-some-random-id-2"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ask-boolean given started when asking question renders 1`] = `
<body>
  <div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    >
      <div
        class="Animate opacity notification flex info enter"
        style="--enter-duration: 100ms; --leave-duration: 100ms;"
      >
        <div
          class="box"
        >
          <i
            class="Icon material focusable"
          >
            <span
              class="icon"
              data-icon-name="info_outline"
            >
              info_outline
            </span>
          </i>
        </div>
        <div
          class="message box grow"
        >
          <div
            class="flex column gaps"
            data-testid="ask-boolean-some-random-id-1"
          >
            <b>
              some-title
            </b>
            <p>
              Some question
            </p>
            <div
              class="flex gaps row align-left box grow"
            >
              <button
                class="Button light"
                data-testid="ask-boolean-some-random-id-1-button-yes"
                type="button"
              >
                Yes
              </button>
              <button
                class="Button active outlined"
                data-testid="ask-boolean-some-random-id-1-button-no"
                type="button"
              >
                No
              </button>
            </div>
          </div>
        </div>
        <div
          class="box"
        >
          <i
            class="Icon close material interactive focusable"
            data-testid="close-notification-for-ask-boolean-for-some-random-id-1"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="close"
            >
              close
            </span>
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ask-boolean given started when asking question when user answers "no" renders 1`] = `
<body>
  <div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;

exports[`ask-boolean given started when asking question when user answers "yes" renders 1`] = `
<body>
  <div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;

exports[`ask-boolean given started when asking question when user closes notification without answering the question renders 1`] = `
<body>
  <div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;
